language/stdlib/modules/DesignatedDealer.move

```diff
@@ -62,69 +64,81 @@ module DesignatedDealer {
+        // modifies global<Dealer>@dd_addr.tiers;
+        ensures len(global<Dealer>(dd_addr).tiers) == len(old(global<Dealer>(dd_addr)).tiers) + 1;
+        ensures global<Dealer>(dd_addr).tiers[len(global<Dealer>(dd_addr).tiers) - 1] == tier_upperbound;
+    }
+
     fun update_tier_(dealer: &mut Dealer, tier_index: u64, new_upperbound: u64) {
         let tiers = &mut dealer.tiers;
         let number_of_tiers = Vector::length(tiers);
-        assert(tier_index <= 3, EINVALID_TIER_INDEX); // max 4 tiers allowed
         assert(tier_index < number_of_tiers, EINVALID_TIER_INDEX);
         // Make sure that this new start for the tier is consistent
-        // with the tier above it.
-        let next_tier = tier_index + 1;
-        if (next_tier < number_of_tiers) {
-            assert(new_upperbound < *Vector::borrow(tiers, next_tier), EINVALID_TIER_START);
+        // with the tier above and below it.
+        let tier = Vector::borrow(tiers, tier_index);
+        if (*tier == new_upperbound) return;
+        if (*tier < new_upperbound) {
+            let next_tier_index = tier_index + 1;
+            if (next_tier_index < number_of_tiers) {
+                assert(new_upperbound < *Vector::borrow(tiers, next_tier_index), EINVALID_TIER_START);
+            };
+        };
+        if (*tier > new_upperbound && tier_index > 0) {
+            let prev_tier_index = tier_index - 1;
+            assert(new_upperbound > *Vector::borrow(tiers, prev_tier_index), EINVALID_TIER_START);
         };
-        let tier_mut = Vector::borrow_mut(tiers, tier_index);
-        *tier_mut = new_upperbound;
+        *Vector::borrow_mut(tiers, tier_index) = new_upperbound;
     }
 
     public fun update_tier(
         tc_account: &signer,
-        addr: address,
+        dd_addr: address,
         tier_index: u64,
         new_upperbound: u64
     ) acquires Dealer {
         assert(Roles::has_treasury_compliance_role(tc_account), EACCOUNT_NOT_TREASURY_COMPLIANCE);
-        let dealer = borrow_global_mut<Dealer>(addr);
+        let dealer = borrow_global_mut<Dealer>(dd_addr);
         update_tier_(dealer, tier_index, new_upperbound)
     }
 
-    fun tiered_mint_(dealer: &mut Dealer, amount: u64, tier_index: u64): bool {
-        // if tier is 4, can mint unlimited
-        assert(tier_index <= 4, EINVALID_TIER_INDEX);
+    spec fun update_tier {
+        // modifies global<Dealer>@dd_addr.tiers;
+        ensures len(global<Dealer>(dd_addr).tiers) == len(old(global<Dealer>(dd_addr)).tiers);
+        ensures global<Dealer>(dd_addr).tiers[tier_index] == new_upperbound;
+    }
+
+    fun tiered_mint_(dealer: &mut Dealer, amount: u64, tier_index: u64) {
         reset_window(dealer);
-        let cur_inflow = *&dealer.window_inflow;
+        let cur_inflow = dealer.window_inflow;
+        let new_inflow = cur_inflow + amount;
         let tiers = &mut dealer.tiers;
         // If the tier_index is one past the bounded tiers, minting is unbounded
```

Comment:
I think we should remove this comment, since line 119 removal (correctly) removes this functionality